src/GraphQL/Extensions/GraphQLBuilderExtensions.cs

```diff
@@ -1172,5 +1172,28 @@ public static IGraphQLBuilder AddExecutionStrategy<TExecutionStrategy>(this IGra
             return builder;
         }
         #endregion
+
+        #region - UseTelemetry -
+#if NET5_0_OR_GREATER
+        /// <summary>
+        /// Configures the GraphQL engine to collect traces via the <see cref="System.Diagnostics.Activity">System.Diagnostics.Activity API</see> and records events that match the
+        /// <see href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/graphql.md">OpenTelemetry recommendations</see>.
+        /// Trace data contains the GraphQL operation name, the operation type, and the optionally the document.
+        /// </summary>
+        /// <remarks>
+        /// When applicable, place after calls to UseAutomaticPersistedQueries to ensure that the query document is recorded properly.
+        /// </remarks>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, Action<GraphQLTelemetryOptions>? configure = null)
```

Comment:
Sorry for late comments, but I have back here after some time.
For now, you are enabling only Traces in `UseTelemetry`. For now, there is no direct support for logs and metrics.

Please consider if it will fits potential extensions for Metrics or you preferer separate methods. like UseTraces/UseActivities and UseMetrics.

Reply:
We used to have `UseMetrics` but this was for Apollo Tracing.  We renamed the method `UseApolloTracing` and this method uses the `Metrics` property from `IResolveFieldContext`.  We can still add `UseLogging` for use with the `ILogging` interface, but this would be a separate NuGet package because the necessary interfaces are not part of .NET.  I think the naming pattern is good.

Reply:
I think he is talking about tracing/metrics/logging as part of OpenTelementry. If you keep the `UseTelemetry` method name, you will have to add metrics and logs support into this method if you decide to add them later. Alternativly, you can rename it to `UseTraces` and removing this coupling.

Reply:
I think for now the goal is to implement the portion that is available in .NET 5+, and nothing specific to the OpenTelemetry library itself. Is there something else I should be aware of that fits this criteria?

Reply:
Well... This is like a situation where you want to swallow 1 large piece in one bite. @Kielek of course we _can_ add support for metrics/traces/logs but it requires more time for design questions. OTel spec indeed provides 3 core concepts - traces, logs, metrics and we _can_ work to integrate somehow all of them. For now this PR aims only traces and only in very limited way - just wrapping graphql execution pipeline into tracing middleware. At the same time, it is worth remembering that we do not want to bring new dependencies to the codebase. Therefore, IMO we need to come up with some flexible API, which will elegantly add new features without breaking changes as implementation evolves. I think we can go along with the same approach as in OpenTelemetry .NET lib - fluent builders:

```
 public static TelemetryBuilder Add[Use?]Telemetry(this IGraphQLBuilder builder, Action<GraphQLTelemetryOptions>? configure = null)

// GraphQLTelemetryOptions become some general-purpose options not scoped exclusevely to traces or we can just get rid of it at the root level of fluent configuration

public sealed class TelemetryBuilder  // looks the same as OpenTelemetryBuilder but for GraphQL.NET, no deps
{
}

// So WithTracing/WithMetrics may be implemented as instance methods in TelemetryBuilder or extension methods somewhere outside
```

so all _telemetry_ stuff will be scoped to `TelemetryBuilder`. And once again - design questions arise. Will it be good to build such a system, the only purpose of which will avoid dependencies on OpenTelemetry .NET lib? Will it be easy to maintain and develop it? How to provide extension points to logs and metrics? And in the end, first we need to determine what metrics we _can_ and/or we _want_ to add, and **only then** look at which API is better to provide for this. 

It is possible that first we should think about whether to add metrics in other ways, for example, through `EventSource` and related APIs:
![изображение](https://user-images.githubusercontent.com/21261007/219937106-a33bf31d-62b5-443c-9ee7-c47ffc191914.png)
I found `EventSource` as rather simple concept to integrate into codebase.

Reply:
I understand that if we make a release with current provided _narrow_ functionality (only traces), then with a high probability it will soon be necessary to make breaking changes to integrate OTel-related metrics and logs.

Reply:
I think it may be much more likely that we add a separate NuGet package for "full" OpenTelemetry compatibility with a new extension method called `UseOpenTelemetry()`.  This would have a much more expansive option set for use with the OpenTelemetry specifically and would be compatible with .NET Standard 2.0, not just .NET 5.  It may be likely that the implementation of `UseOpenTelemetry()` would call other existing methods within GraphQL.NET such as `UseTelemetry()` and simply expand on their functionality.

Reply:
Yeah. Hybrid design - core telemetry features in the main package without additional deps + rich telemetry features in a separate package with OpenTelemetry nuget dependency.

Reply:
@gao-artur correctly read my intentions. I have tried to signalize that there potentially should be extended by Logs and Metrics OTel feature. IMO it is unlikely as long you do not include any new dependency to the library (OTel logs requiers ILogger, Metrics - System.Diagnostics.DiagnosticSource).

Reply:
I suggest you open new issue to track progress there.